docs/features/CRAWL_RULES.md

```diff
@@ -0,0 +1,104 @@
+
+### `begins`
+
+- The path pattern is a literal string except for the character `*`, which is a meta character that will match anything
+- The rule matches when the path pattern matches the beginning of the path (which always begins with `/`)
+- If using this rule, begin your path pattern with `/`
+
+### `ends`
+
+- The path pattern is a literal string except for the character `*`, which is a meta character that will match anything
+- The rule matches when the path pattern matches the end of the path
+
+### `contains`
+
+- The path pattern is a literal string except for the character `*`, which is a meta character that will match anything.
+- The rule matches when the path pattern matches anywhere within the path.
+
+### `regex`
+
+- The path pattern is a regular expression compatible with the Ruby language regular expression engine
+  - In addition to literal characters, the path pattern may include:
+    - [Metacharacters](https://ruby-doc.org/core-3.1.0/Regexp.html#class-Regexp-label-Metacharacters+and+Escapes)
+    - [Character classes](https://ruby-doc.org/core-3.1.0/Regexp.html#class-Regexp-label-Character+Classes)
+    - [Repetitions](https://ruby-doc.org/core-3.1.0/Regexp.html#class-Regexp-label-Repetition)
+  - You can test Ruby regular expressions using [Rubular](https://rubular.com/)
+- The rule matches when the path pattern matches the beginning of the path (which always begins with `/`)
+- If using this rule, begin your path pattern with `\/` or a metacharacter or character class that matches `/`
+
+## Crawl rule matching
+
+The following table provides various examples of crawl rule matching:
+
+| URL path                  | Rule type  | Path pattern   | Match? |
+|---------------------------|------------|----------------|--------|
+| `/foo/bar`                | `begins`   | `/foo`         | YES    |
+| `/foo/bar`                | `begins`   | `/*oo`         | YES    |
+| `/bar/foo`                | `begins`   | `/foo`         | NO     |
+| `/foo/bar`                | `begins`   | `foo`          | NO     |
+| `/blog/posts/hello-world` | `ends`     | `world`        | YES    |
+| `/blog/posts/hello-world` | `ends`     | `hello-*`      | YES    |
+| `/blog/world-hello `      | `ends`     | `world `       | NO     |
+| `/blog/world-hello`       | `ends`     | `*world`       | NO     |
+| `/fruits/bananas`         | `contains` | `banana`       | YES    |
+| `/fruits/apples`          | `contains` | `banana`       | NO     |
+| `/2020`                   | `regex`    | `\/[0-9]{3,5}` | YES    |
+| `/20`                     | `regex`    | `\/[0-9]{3,5}` | NO     |
+| `/2020`                   | `regex`    | `[0-9]{3,5}`   | NO     |
+
+## Restricting paths using crawl rules
+
+The example config crawl rules has a default crawl rule for the example.
+Below is a simplified version of that rule:
+
+```yaml
+domains:
+  - url: http://example.com
+  - crawl_rules:
+    - policy: allow
+      type: regex
+      pattern: .*
+```
```

Comment:
Is this rule explicitly required at the end of the list? Or will we make it implicit? Since in Elastic Crawler we could actually enforce this was always there, but now we don't have the same control over the list of rules that are persisted.

Reply:
The rule is explicity required at the end of the list, yes. But only if the user has configured _any_ crawl rules. I think this is where it differs from Elastic Crawler. Elastic Crawler will always have at least one rule defined (`regex` `.*`), even if the user hasn't defined anything. For Open Crawler, users can just entirely omit crawl rules.

I think the question comes down to "should enabling crawl rules implicitly allow or deny all URLs that don't match a rule". I think that's worth discussing more off the PR.

Reply:
For now I'll merge this but we can address this in another PR if we decide to change it.

Reply:
Actually, looking at Elastic Crawler I see your point. Because it's not possible to remove the final catch-all rule, it effectively always allows all URLs that don't match a rule.

I'll change it so users don't need to include this in every configuration.